### ISSUE TITLE
Duplicate items

### ISSUE THREAD

**Issue by Etenil** [SSilence/selfoss#208]
The feed reader sometimes fails to detect the presence of an item within the database an thus creates multiple items.

This only happens with Slashdot's rss feed on my server, and slashdot uses some very long hyperlink as uid, which might be the cause of the problem.

I have been able to solve this on my server by modifying the dao and update code so that items'presence in the database be determined by both their uid and their link. So far it's worked OK, but I don't think it's a clean fix and thus will not commit it.

**Comment by bitvijays** [comment]
@SSilence  Thank you for authoring selfoss. Helps a lot. Also, The issue of duplicates is still existing. I am getting duplicate for Packetstorm News/ Packetstorm Files. Probably, you might want to see the "Link" and "Source" in the database, to see if it's already present?

**Comment by jtojnar** [comment]
@bitvijays could you please open a new issue with some more details? Can you locate the duplicate articles in the database?

**Comment by niol** [comment]
The results of a query such as this one would help us find the right solution to this:
```sql
SELECT items.uid, items2.uid, items.title, items2.title FROM items, items AS items2 WHERE items.source = items2.source AND items.id > items2.id AND items.link = items2.link;
```
I have a feed that provides new items with the same link, so the link alone is not good. Also I'm against avoiding duplicated across multiple feeds because this allows one feed to prevent items from other feeds to get into the db.

**Comment by bitvijays** [comment]
[selfoss.zip](https://github.com/SSilence/selfoss/files/944920/selfoss.zip)

@niol @jtojnar Here's the zip file containing the sqlite database. ( 1 MB only ). Hopefully, this should provide more insight.

**Comment by jtojnar** [comment]
I do not understand how could this happen. On the bright side, I [fixed favicons](https://github.com/SSilence/selfoss/pull/920) for two of your feeds 😉

**Comment by niol** [comment]
Duplicated ```uid```s may only happen if multiple parallel updates are running, for instance if a cronjob is running while a manual update is triggered manually.

I had proposed something to fix this some time ago (see #597) which was using a file lock to prevent concurrent updates on the same source. Another option would be to add a constraint on the ```items``` table ensuring that ```(uid, source)``` is unique and handle the ```INSERT``` error properly.

**Comment by jtojnar** [comment]
@niol I think using `UNIQUE` constraint is preferable due to a lower number of moving parts. Additionally, instead of handling an error, [`UPSERT`](https://en.wikipedia.org/wiki/Merge_(SQL)#Other_non-standard_implementations) can possibly be used, though it is [hairy on PostgreSQL < 9.5](https://stackoverflow.com/questions/17267417/how-to-upsert-merge-insert-on-duplicate-update-in-postgresql)
